File: 0035-search-insert-position/0035-search-insert-position.dart
class Solution {
  int searchInsert(List<int> nums, int target) {
    int l = 0, r = nums.length-1;
    while(l <= r){
        int mid = (r + l) ~/2;
        if(nums[mid] <= target && (mid+1 == nums.length || nums[mid+1] > target)){
            return nums[mid] == target ? mid : mid+1;
        }
        else if(nums[mid] < target){
            l = mid+1;
        }
        else {
            r = mid-1;
        }
    }
    return 0;
  }
}

File: 0035-search-insert-position/0035-search-insert-position.java
class Solution {
    public int searchInsert(int[] nums, int target) {
        int left = 0, right = nums.length-1;
        if(nums[left] >= target) return left;
        else if(nums[right] <= target) return nums[right] == target ? right : right+1;
        while(left <= right){
            int mid = (right + left) / 2;
            if(nums[mid] == target) return mid;
            else if(nums[mid] < target && target < nums[mid+1]) return mid+1;
            else if(nums[mid] > target){
                right = mid - 1;
            }else {
                left = mid + 1;
            }
        }
        return left;
    }
}